File: defaultcontent/Template.java
// @BASEFILENAME@
// ------------------------------------------------------------------
//
// Description goes here....
//
// Author: @AUTHOR@
// Created @DATE@
//
// Last saved: <2014-March-19 22:45:11>
// ------------------------------------------------------------------
//
@INSERTFILE(/Users/Dino/elisp/defaultcontent/Copyright.txt)@
//
// ------------------------------------------------------------------

package com.ionic.whatever;

import java.util.Hashtable;
import java.util.Enumeration;

public class @BASEFILENAMELESSEXTENSION@ {
    private final String optString = "vu:p:s:"; // getopt style

    // public @BASEFILENAMELESSEXTENSION@ () {} // uncomment if wanted

    public @BASEFILENAMELESSEXTENSION@ (String[] args)
        throws java.lang.Exception {
        GetOpts(args, optString);
    }

    private Hashtable<String, Object> options = new Hashtable<String, Object> ();

    private void GetOpts(String[] args, String optString)
        throws java.lang.Exception {
        // Parse command line args for args in the following format:
        //   -a value -b value2 ... ...

        // sanity checks
        if (args == null) return;
        if (args.length == 0) return;
        if (optString == null) return;
        final String argPrefix = "-";
        String patternString = "^" + argPrefix + "([" + optString.replaceAll(":","") + "])";

        java.util.regex.Pattern p = java.util.regex.Pattern.compile(patternString);

        int L = args.length;
        for(int i=0; i < L; i++) {
            String arg = args[i];
            java.util.regex.Matcher m = p.matcher(arg);
            if (!m.matches()) {
                throw new java.lang.Exception("The command line arguments are improperly formed. Use a form like '-a value' or just '-b' .");
            }

            char ch = arg.charAt(1);
            int pos = optString.indexOf(ch);

            if ((pos != optString.length() - 1) && (optString.charAt(pos+1) == ':')) {
                if (i+1 < L) {
                    i++;
                    Object current = this.options.get(m.group(1));
                    ArrayList<String> newList;
                    if (current == null) {
                        // not a previously-seen option
                        this.options.put(m.group(1), args[i]);
                    }
                    else if (current instanceof ArrayList<?>) {
                        // previously seen, and already a lsit
                        newList = (ArrayList<String>) current;
                        newList.add(args[i]);
                    }
                    else {
                        // we have one value, need to make a list
                        newList = new ArrayList<String>();
                        newList.add((String)current);
                        newList.add(args[i]);
                        this.options.put(m.group(1), newList);
                    }
                }
                else {
                    throw new java.lang.Exception("Incorrect arguments.");
                }
            }
            else {
                // a "no-value" argument, like -v for verbose
                options.put(m.group(1), (Boolean) true);
            }
        }
    }

    private static ArrayList<String> asOptionsList(Object o) {
        if (o instanceof ArrayList<?>) {
            return (ArrayList<String>) o;
        }

        ArrayList<String> list = new ArrayList<String>();

        if (o instanceof String) {
            list.add((String)o);

        }
        return list;
    }

    private void maybeShowOptions() {
        Boolean verbose = (Boolean) this.options.get("v");
        if (verbose != null && verbose) {
            System.out.println("options:");
            Enumeration e = this.options.keys();
            while(e.hasMoreElements()) {
                // iterate through Hashtable keys Enumeration
                String k = (String) e.nextElement();
                Object o = this.options.get(k);
                String v = null;
                v = (o.getClass().equals(Boolean.class)) ?  "true" : (String) o;
                System.out.println("  " + k + ": " + v);
            }

            // enumerate properties here?
        }
    }


    public void Run() {

        maybeShowOptions();
        Boolean verbose = (Boolean) this.options.get("v");

        // for repeated options:
        ArrayList<String> prefixes = asOptionsList(Options.get("n"));
        if (prefixes != null) {
            ...
        }

        @DOT@
    }


    public static void Usage() {
        System.out.println("@BASEFILENAMELESSEXTENSION@: <usage statement here>.\n");
        System.out.println("Usage:\n  java @BASEFILENAMELESSEXTENSION@ [-v]  -f <value>");
    }


    public static void main(String[] args) {
        try {
            @BASEFILENAMELESSEXTENSION@ me = new @BASEFILENAMELESSEXTENSION@(args);
            me.Run();
        }
        catch (java.lang.Exception exc1) {
            System.out.println("Exception:" + exc1.toString());
            exc1.printStackTrace();
        }
    }

}

File: defaultcontent/Template.cs
// @BASEFILENAME@
// ------------------------------------------------------------------
//
// Description goes here....
//
// Author: @AUTHOR@
// created on host: @ENV(COMPUTERNAME)@
// initialized on: @DATE@
//
// last saved: <2012-March-24 02:08:46>
// ------------------------------------------------------------------
//
@INSERTFILE(c:/Users/Dino/elisp/defaultcontent/Copyright.txt)@//
// ------------------------------------------------------------------

using System;
using System.Reflection;


// to allow fast ngen
[assembly: AssemblyTitle("@BASEFILENAME@")]
[assembly: AssemblyDescription("insert purpose here")]
[assembly: AssemblyConfiguration("")]
[assembly: AssemblyCompany("Dino Chiesa")]
[assembly: AssemblyProduct("Tools")]
[assembly: AssemblyCopyright("Copyright (c) Dino Chiesa 2010")]
[assembly: AssemblyTrademark("")]
[assembly: AssemblyCulture("")]
[assembly: AssemblyVersion("1.1.1.1")]


namespace Ionic.ToolsAndTests
{

  public class @BASEFILENAMELESSEXTENSION@
  {
    // ctor
    public @BASEFILENAMELESSEXTENSION@ () {}

    string _positionalParam1;
    string _param1;
    int _intParam = DefaultIntParamValue;
    bool _flag1;
    int _size;

    private readonly int DefaultIntParamValue = -99;

    // ctor
    public @BASEFILENAMELESSEXTENSION@ (string[] args)
    {
        for (int i=0; i < args.Length; i++)
        {
            switch (args[i])
            {
                case "-stringArg":
                    i++;
                    if (args.Length <= i) throw new ArgumentException(args[i]);
                    _param1 = args[i];
                    break;

                case "-s":
                    i++;
                    if (args.Length <= i) throw new Exception(args[i-1]);
                    if (args[i].ToUpper().EndsWith("K"))
                        _size = System.Int32.Parse(args[i].Substring(0,args[i].Length-1)) * 1024;
                    else if (args[i].ToUpper().EndsWith("KB"))
                        _size = System.Int32.Parse(args[i].Substring(0,args[i].Length-2)) * 1024;
                    else if (args[i].ToUpper().EndsWith("M"))
                        _size = System.Int32.Parse(args[i].Substring(0,args[i].Length-1)) * 1024*1024;
                    else if (args[i].ToUpper().EndsWith("MB"))
                        _size = System.Int32.Parse(args[i].Substring(0,args[i].Length-2)) * 1024*1024;
                    else
                        _size = Int32.Parse(args[i]);
                    break;


                case "-intArg":
                    i++;
                    if (args.Length <= i) throw new ArgumentException(args[i]);
                    if (_intParam != DefaultIntParamValue)
                        throw new ArgumentException(args[i]);
                    if (args[i].StartsWith("0x"))
                        _intParam = System.Int32.Parse(args[i].Substring(2), System.Globalization.NumberStyles.AllowHexSpecifier );
                    else
                        _intParam = System.Int32.Parse(args[i]);
                    break;


                case "-boolarg":
                    _flag1 = true;
                    break;

                case "-?":
                    throw new ArgumentException(args[i]);

                default:
                    if (_positionalParam1 != null)
                        throw new ArgumentException(args[i]);

                    _positionalParam1 = args[i];
                    break;
            }
        }

        // default values
        if (_positionalParam1 == null)
            _positionalParam1 = "default.value.for.positional.param";

        if (_param1 == null)
            _param1 = "default.value.for.param1";

        if (_param2 == 0)
            _param2 = DEFAULT_value_for_param2;

    }

    public void Run()
    {

        @DOT@

    }

    public static void Usage()
    {
        Console.WriteLine("\n@BASEFILENAMELESSEXTENSION@: <usage statement here>.\n");
        Console.WriteLine("Usage:\n  @BASEFILENAMELESSEXTENSION@ [-arg1 <value>] [-arg2]");
    }


    public static void Main(string[] args)
    {
      try
      {
        new @BASEFILENAMELESSEXTENSION@(args)
            .Run();
      }
      catch (System.Exception exc1)
      {
        Console.WriteLine("Exception: {0}", exc1.ToString());
        Usage();
      }
    }

  }
}


